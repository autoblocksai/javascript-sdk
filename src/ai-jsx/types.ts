import type { Node } from 'ai-jsx';

export interface AutoblocksPlaceholderProps {
  children: Node;
  name: string;
}

export interface ABSpan {
  id: string;
  parentId: string | undefined;
  memoizedId: string | undefined;
  name: string;
  props: unknown;
  startTime: string;
  endTime: string | undefined;
  children: (ABSpan | string)[];
}
