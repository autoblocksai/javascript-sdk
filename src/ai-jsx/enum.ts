export enum AutoblocksLoggerAttribute {
  PROMPT_TRACKING = 'autoblocks.promptTracking',
}
