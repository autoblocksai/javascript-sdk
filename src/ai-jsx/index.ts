export { AutoblocksJsxTracer } from './tracer';
export { AutoblocksPlaceholder } from './util';
