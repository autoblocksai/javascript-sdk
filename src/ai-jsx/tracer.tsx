import { AsyncLocalStorage } from 'node:async_hooks';
import * as AIJSX from 'ai-jsx';
import crypto from 'crypto';
import {
  findMemoizedId,
  bindAsyncGenerator,
  processCompletedRootSpan,
  makeComponentName,
} from './util';
import type { AutoblocksSpan } from './types';

export function AutoblocksJsxTracer(
  props: {
    children: AIJSX.Node;
  },
  { wrapRender }: AIJSX.ComponentContext,
) {
  const currentSpanStorage = new AsyncLocalStorage<AutoblocksSpan>();

  return AIJSX.withContext(
    <>{props.children}</>,
    wrapRender(
      (render) => (renderContext, renderable, shouldStop, appendOnly) => {
        const parentSpan = currentSpanStorage.getStore();

        if (!AIJSX.isElement(renderable)) {
          if (
            parentSpan &&
            ['string', 'boolean', 'number'].includes(typeof renderable)
          ) {
            parentSpan.children.push(`${renderable}`);
          }

          return render(renderContext, renderable, shouldStop, appendOnly);
        }

        const newSpan = {
          id: crypto.randomUUID(),
          parentId: parentSpan?.id,
          memoizedId: findMemoizedId(renderable.props),
          name: makeComponentName(renderable.tag),
          props: renderable.props,
          startTime: new Date().toISOString(),
          endTime: undefined,
          children: [],
        };

        if (parentSpan) {
          parentSpan.children.push(newSpan);
        }

        return currentSpanStorage.run(newSpan, () => {
          async function* gen() {
            const currentSpan = currentSpanStorage.getStore();
            try {
              return yield* render(
                renderContext,
                renderable,
                shouldStop,
                appendOnly,
              );
            } finally {
              if (currentSpan) {
                currentSpan.endTime = new Date().toISOString();
                if (!parentSpan) {
                  // The current span is the root span and is complete
                  try {
                    await processCompletedRootSpan(currentSpan);
                  } catch (e) {
                    console.error(`Error processing completed span: ${e}`);
                  }
                }
              }
            }
          }

          return bindAsyncGenerator(gen());
        });
      },
    ),
  );
}
