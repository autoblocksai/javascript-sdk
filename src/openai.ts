import crypto from 'crypto';
import { AutoblocksTracer } from './tracer';
import { readEnv, AUTOBLOCKS_INGESTION_KEY } from './util';

const tracer = new AutoblocksTracer(readEnv(AUTOBLOCKS_INGESTION_KEY) || '', {
  properties: { provider: 'openai' },
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function makeWrapper(func: any): any {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return async function wrapper(this: any, ...args: any[]): Promise<any> {
    const original = func.bind(this);

    let error = undefined;
    let response = undefined;

    let traceId = undefined;
    if (!tracer.traceId) {
      // Note we don't use setTraceId but instead pass the traceId with each sendEvent call
      // so that we can tell when a user has set a traceId via setTraceId. If we were to call
      // setTraceId here, we would not be able to tell it was set by us vs. set by a user elsewhere.
      traceId = crypto.randomUUID();
    }

    await tracer.sendEvent('ai.completion.request', {
      traceId,
      properties: args[0],
    });

    const start = Date.now();

    try {
      response = await original(...args);
      return response;
    } catch (err) {
      error = err;
      throw err;
    } finally {
      const latencyMs = Date.now() - start;

      if (error) {
        await tracer.sendEvent('ai.completion.error', {
          traceId,
          properties: {
            latencyMs,
            error: error.toString(),
          },
        });
      } else {
        await tracer.sendEvent('ai.completion.response', {
          traceId,
          properties: {
            latencyMs,
            response,
          },
        });
      }
    }
  };
}

async function patch() {
  const openAiModule = await import('openai');
  openAiModule.OpenAI.Completions.prototype.create = makeWrapper(
    openAiModule.OpenAI.Completions.prototype.create,
  );
  openAiModule.OpenAI.Chat.Completions.prototype.create = makeWrapper(
    openAiModule.OpenAI.Chat.Completions.prototype.create,
  );
}

export async function traceOpenAI(): Promise<AutoblocksTracer> {
  if (traceOpenAI.called) {
    return tracer;
  }

  if (!readEnv(AUTOBLOCKS_INGESTION_KEY)) {
    throw new Error(
      `You must set the ${AUTOBLOCKS_INGESTION_KEY} environment variable in order to use traceOpenAI.`,
    );
  }

  try {
    await patch();
  } catch (err) {
    console.warn(`Couldn't patch openai module: ${err}`);
  }

  traceOpenAI.called = true;

  return tracer;
}

traceOpenAI.called = false;
