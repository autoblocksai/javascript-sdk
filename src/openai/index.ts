import crypto from 'crypto';
import { AutoblocksTracer } from '../tracer';
import { readEnv, AUTOBLOCKS_INGESTION_KEY } from '../util';

let tracer: AutoblocksTracer | undefined = undefined;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function makeWrapper(func: any): any {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return async function wrapper(this: any, ...args: any[]): Promise<any> {
    const original = func.bind(this);

    if (!tracer) {
      return original(...args);
    }

    let error = undefined;
    let response = undefined;

    let traceId = undefined;
    if (!tracer.traceId) {
      // Note we don't use setTraceId but instead pass the traceId with each sendEvent call
      // so that we can tell when a user has set a traceId via setTraceId. If we were to call
      // setTraceId here, we would not be able to tell it was set by us vs. set by a user elsewhere.
      traceId = crypto.randomUUID();
    }

    // Similar to the comment above, we don't set this via tracer.updateProperties
    // because the tracer is returned from this function and meant to be used by the
    // end user for additional events, and those events should not belong to this span.
    const spanId = crypto.randomUUID();

    await tracer.sendEvent('ai.completion.request', {
      traceId,
      spanId,
      properties: args[0],
    });

    const start = Date.now();

    try {
      response = await original(...args);
      return response;
    } catch (err) {
      error = err;
      throw err;
    } finally {
      const latencyMs = Date.now() - start;

      if (error) {
        await tracer.sendEvent('ai.completion.error', {
          traceId,
          spanId,
          properties: {
            latencyMs,
            error: error.toString(),
          },
        });
      } else {
        await tracer.sendEvent('ai.completion.response', {
          traceId,
          spanId,
          properties: {
            latencyMs,
            response,
          },
        });
      }
    }
  };
}

async function patch() {
  const openAiModule = await import('openai');
  openAiModule.OpenAI.Completions.prototype.create = makeWrapper(
    openAiModule.OpenAI.Completions.prototype.create,
  );
  openAiModule.OpenAI.Chat.Completions.prototype.create = makeWrapper(
    openAiModule.OpenAI.Chat.Completions.prototype.create,
  );
}

export async function traceOpenAI(): Promise<AutoblocksTracer> {
  if (tracer) {
    return tracer;
  }

  const ingestionKey = readEnv(AUTOBLOCKS_INGESTION_KEY);
  if (!ingestionKey) {
    throw new Error(
      `You must set the ${AUTOBLOCKS_INGESTION_KEY} environment variable in order to use traceOpenAI.`,
    );
  }
  tracer = new AutoblocksTracer({
    ingestionKey,
    properties: {
      provider: 'openai',
    },
  });

  try {
    await patch();
  } catch (err) {
    console.warn(`Couldn't patch openai module: ${err}`);
  }

  return tracer;
}
