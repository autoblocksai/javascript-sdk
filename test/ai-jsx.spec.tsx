import * as AI from 'ai-jsx';
import {
  ChatCompletion,
  UserMessage,
  SystemMessage,
} from 'ai-jsx/core/completion';
import { AnthropicChatModel } from 'ai-jsx/lib/anthropic';
import {
  AutoblocksJsxTracer,
  AutoblocksLoggerAttribute,
  AutoblocksPlaceholder,
  AutoblocksTemplateSelect,
  AutoblocksTemplateSelectItem,
} from '../src/ai-jsx/index';
import { OpenAIChatModel } from 'ai-jsx/lib/openai';
import { LogImplementation } from 'ai-jsx/core/log';

interface SentEvent {
  message: string;
  traceId: string;
  timestamp: string;
  properties: {
    spanId: string;
    parentSpanId?: string;
    latency?: number;
    [key: string]: unknown;
  };
}

jest.setTimeout(100000);

const expectAllEqualAndDefined = (xs: string[]) => {
  expect(xs.every((x) => x === xs[0])).toBe(true);
  expect(xs.every(Boolean)).toBe(true);
};

describe('ai-jsx', () => {
  beforeAll(() => {
    process.env.AUTOBLOCKS_INGESTION_KEY = 'test';
  });

  afterAll(() => {
    process.env.AUTOBLOCKS_INGESTION_KEY = undefined;
  });

  let mockFetch: jest.SpyInstance;
  const originalFetch = global.fetch;

  beforeEach(() => {
    mockFetch = jest
      .spyOn(global, 'fetch')
      .mockImplementation((...args) => originalFetch(...args));
  });

  const sentEvents = () =>
    mockFetch.mock.calls
      .filter((c) => c[0].startsWith('https://ingest-event.autoblocks.ai'))
      .map((c) => JSON.parse(c[1].body)) as SentEvent[];

  const makeSpanPairs = (events: SentEvent[]) => {
    const pairs: Record<string, SentEvent[]> = {};
    for (const event of events) {
      if (!pairs[event.properties.spanId]) {
        pairs[event.properties.spanId] = [];
      }
      pairs[event.properties.spanId].push(event);
    }
    return pairs;
  };

  // This needs to return a function b/c ???
  const makeAssertions = () => () => {
    const events = sentEvents();

    // All events should have the same trace id
    expectAllEqualAndDefined(events.map((r) => r.traceId));

    // There should be an even amount of events since each completion
    // sends a request and response event
    expect(events.length).toBeGreaterThan(0);
    expect(events.length % 2).toBe(0);

    const pairs = makeSpanPairs(events);
    expect(Object.keys(pairs).length).toEqual(events.length / 2);

    for (const pair of Object.values(pairs)) {
      expect(pair.length).toEqual(2);
      expect(pair[0].message).toEqual('ai.completion.request');
      expect(
        ['ai.completion.response', 'ai.completion.error'].includes(
          pair[1].message,
        ),
      ).toBe(true);
      expect(pair[0].timestamp < pair[1].timestamp).toBe(true);
      expect(pair[1].properties.latency).toBeDefined();
      expect(pair[1].properties.latency).toBeGreaterThan(0);
    }

    for (const request of events) {
      const matchers = {
        traceId: expect.any(String),
        timestamp: expect.any(String),
        properties: {
          latency: expect.any(Number),
          spanId: expect.any(String),
          parentSpanId: expect.any(String),
        },
      };

      if (request.properties.parentSpanId === undefined) {
        delete matchers.properties.parentSpanId;
      }
      if (request.properties.latency === undefined) {
        delete matchers.properties.latency;
      }

      expect(request).toMatchSnapshot(matchers);
    }
  };

  it('works with no tracker id', async () => {
    await AI.createRenderContext().render(
      <AutoblocksJsxTracer>
        <ChatCompletion temperature={0} model="gpt-3.5-turbo">
          <SystemMessage>
            You are a helpful assistant. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>What color is the sky?</UserMessage>
        </ChatCompletion>
      </AutoblocksJsxTracer>,
    );

    makeAssertions()();
  });

  it('handles placeholders', async () => {
    await AI.createRenderContext().render(
      <AutoblocksJsxTracer>
        <ChatCompletion
          temperature={0}
          model="gpt-3.5-turbo"
          autoblocks-tracker-id="my-tracker-id"
        >
          <SystemMessage>
            You are a helpful assistant. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What color is the{' '}
            <AutoblocksPlaceholder name="thing">sky</AutoblocksPlaceholder>?
          </UserMessage>
        </ChatCompletion>
      </AutoblocksJsxTracer>,
    );

    makeAssertions()();
  });

  it('handles selects', async () => {
    enum Select {
      A = 'A',
      B = 'B',
    }

    const runtime = () => 'runtime';

    await AI.createRenderContext().render(
      <AutoblocksJsxTracer>
        <ChatCompletion
          temperature={0}
          model="gpt-3.5-turbo"
          autoblocks-tracker-id="my-tracker-id"
        >
          <SystemMessage>
            <AutoblocksTemplateSelect
              name="instructions"
              selectedItemName={Select.A}
            >
              <AutoblocksTemplateSelectItem name={Select.A}>
                You are a helpful assistant. Always respond with one word in all
                lowercase letters and no punctuation.
              </AutoblocksTemplateSelectItem>
              <AutoblocksTemplateSelectItem name={Select.B}>
                <>
                  This template won't be chosen at {runtime()}, but it should
                  still be included in the list of{' '}
                  <AutoblocksPlaceholder name="templates">
                    templates
                  </AutoblocksPlaceholder>
                  .
                </>
              </AutoblocksTemplateSelectItem>
            </AutoblocksTemplateSelect>
          </SystemMessage>
          <UserMessage>
            What color is the{' '}
            <AutoblocksPlaceholder name="thing">sky</AutoblocksPlaceholder>?
          </UserMessage>
        </ChatCompletion>
      </AutoblocksJsxTracer>,
    );

    makeAssertions()();
  });

  it('handles nested completions', async () => {
    const GetThingColor = (props: { thing: string }) => {
      return (
        <ChatCompletion
          temperature={0}
          model="gpt-3.5-turbo"
          autoblocks-tracker-id="get-thing-color"
        >
          <SystemMessage>
            You are a helpful assistant. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What color is the{' '}
            <AutoblocksPlaceholder name="thing">
              {props.thing}
            </AutoblocksPlaceholder>
            ?
          </UserMessage>
        </ChatCompletion>
      );
    };

    await AI.createRenderContext().render(
      <AutoblocksJsxTracer>
        <ChatCompletion
          temperature={0}
          model="gpt-3.5-turbo"
          autoblocks-tracker-id="determine-color-combination"
        >
          <SystemMessage>
            You are an expert in colors. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What do you get when you mix red with <GetThingColor thing="sky" />?
          </UserMessage>
        </ChatCompletion>
      </AutoblocksJsxTracer>,
    );

    const events = sentEvents();
    expect(events.map((e) => e.message)).toEqual([
      'ai.completion.request',
      'ai.completion.response',
      'ai.completion.request',
      'ai.completion.response',
    ]);

    expect(events.map((e) => e.properties.spanId)).toEqual([
      events[0].properties.spanId,
      events[0].properties.spanId,
      events[2].properties.spanId,
      events[2].properties.spanId,
    ]);

    expect(events.map((e) => e.properties.parentSpanId)).toEqual([
      undefined,
      undefined,
      events[0].properties.spanId,
      events[0].properties.spanId,
    ]);

    makeAssertions()();
  });

  it('handles nested completions w/o tracking IDs', async () => {
    const GetThingColor = (props: { thing: string }) => {
      return (
        <ChatCompletion temperature={0} model="gpt-3.5-turbo">
          <SystemMessage>
            You are a helpful assistant. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What color is the{' '}
            <AutoblocksPlaceholder name="thing">
              {props.thing}
            </AutoblocksPlaceholder>
            ?
          </UserMessage>
        </ChatCompletion>
      );
    };

    await AI.createRenderContext().render(
      <AutoblocksJsxTracer>
        <ChatCompletion temperature={0} model="gpt-3.5-turbo">
          <SystemMessage>
            You are an expert in colors. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What do you get when you mix red with <GetThingColor thing="sky" />?
          </UserMessage>
        </ChatCompletion>
      </AutoblocksJsxTracer>,
    );

    const events = sentEvents();
    expect(events.map((e) => e.message)).toEqual([
      'ai.completion.request',
      'ai.completion.response',
      'ai.completion.request',
      'ai.completion.response',
    ]);

    expect(events.map((e) => e.properties.spanId)).toEqual([
      events[0].properties.spanId,
      events[0].properties.spanId,
      events[2].properties.spanId,
      events[2].properties.spanId,
    ]);

    expect(events.map((e) => e.properties.parentSpanId)).toEqual([
      undefined,
      undefined,
      events[0].properties.spanId,
      events[0].properties.spanId,
    ]);

    makeAssertions()();
  });

  it('handles anthropic chat models', async () => {
    await AI.createRenderContext().render(
      <AutoblocksJsxTracer>
        <AnthropicChatModel temperature={0} model="claude-instant-1">
          <SystemMessage>
            You are a helpful assistant. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>What color is the sky?</UserMessage>
        </AnthropicChatModel>
      </AutoblocksJsxTracer>,
    );

    makeAssertions()();
  });

  it('handles errors', async () => {
    try {
      await AI.createRenderContext().render(
        <AutoblocksJsxTracer>
          <ChatCompletion temperature={0} model="unsupported-model">
            <SystemMessage>
              You are a helpful assistant. Always respond with one word in all
              lowercase letters and no punctuation.
            </SystemMessage>
            <UserMessage>What color is the sky?</UserMessage>
          </ChatCompletion>
        </AutoblocksJsxTracer>,
      );
    } catch {
      // expected
    }

    makeAssertions()();
  });

  it('sets prompt tracking attributes on the logger', async () => {
    const mockSetAttribute = jest.fn();

    class MockLogger extends LogImplementation {
      log() {}

      setAttribute(
        _element: AI.Element<any>,
        _renderId: string,
        _key: string,
        _value: string,
      ): void {
        mockSetAttribute(_element, _renderId, _key, _value);
      }
    }

    const MyCustomChatModel = OpenAIChatModel;

    const GetThingColor = (props: { thing: string }) => {
      return (
        <MyCustomChatModel
          temperature={0}
          model="gpt-3.5-turbo"
          autoblocks-tracker-id="get-thing-color"
        >
          <SystemMessage>
            You are a helpful assistant. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What color is the{' '}
            <AutoblocksPlaceholder name="thing">
              {props.thing}
            </AutoblocksPlaceholder>
            ?
          </UserMessage>
        </MyCustomChatModel>
      );
    };

    const GetColorCombo = (props: { color: string; thing: string }) => {
      return (
        <MyCustomChatModel
          temperature={0}
          model="gpt-3.5-turbo"
          autoblocks-tracker-id="determine-color-combination"
        >
          <SystemMessage>
            You are an expert in colors. Always respond with one word in all
            lowercase letters and no punctuation.
          </SystemMessage>
          <UserMessage>
            What do you get when you mix{' '}
            <AutoblocksPlaceholder name="color">
              {props.color}
            </AutoblocksPlaceholder>{' '}
            with <GetThingColor thing={props.thing} />?
          </UserMessage>
        </MyCustomChatModel>
      );
    };

    await AI.createRenderContext({ logger: new MockLogger() }).render(
      <AutoblocksJsxTracer
        skipSendingEvents={true}
        customChatModelComponent={MyCustomChatModel}
      >
        <GetColorCombo color="red" thing="sky" />
      </AutoblocksJsxTracer>,
    );

    const promptTrackingCalls = mockSetAttribute.mock.calls.filter(
      (c) => c[2] === AutoblocksLoggerAttribute.PROMPT_TRACKING,
    );

    // There should be two prompt tracking setAttribute calls,
    // one for each chat model component with an autoblocks-tracker-id
    // property
    expect(promptTrackingCalls.length).toEqual(2);

    // The render IDs should both be defined and be different
    const renderIds = promptTrackingCalls.map((c) => c[1]);
    expect(renderIds.every(Boolean)).toBe(true);
    expect(new Set(renderIds).size).toEqual(2);

    // We should have logged the templates for each chat model component
    expect(JSON.parse(promptTrackingCalls[0][3])).toEqual({
      id: 'get-thing-color',
      templates: [
        {
          id: 'get-thing-color/system',
          template:
            'You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.',
        },
        {
          id: 'get-thing-color/user',
          template: 'What color is the {{ thing }}?',
        },
      ],
    });

    expect(JSON.parse(promptTrackingCalls[1][3])).toEqual({
      id: 'determine-color-combination',
      templates: [
        {
          id: 'determine-color-combination/system',
          template:
            'You are an expert in colors. Always respond with one word in all lowercase letters and no punctuation.',
        },
        {
          id: 'determine-color-combination/user',
          template:
            'What do you get when you mix {{ color }} with {{ get-thing-color }}?',
        },
      ],
    });
  });
});
