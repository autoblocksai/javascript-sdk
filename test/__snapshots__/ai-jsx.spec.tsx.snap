// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ai-jsx handles anthropic chat models 1`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "What color is the sky?",
        "role": "user",
      },
      {
        "content": "For subsequent replies you will adhere to the following instructions: You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "user",
      },
      {
        "content": "Okay, I will do that.",
        "role": "assistant",
      },
    ],
    "model": "claude-instant-1",
    "provider": "anthropic",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles anthropic chat models 2`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "blue",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "provider": "anthropic",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 2,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles errors 1`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
    ],
    "model": "unsupported-model",
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles errors 2`] = `
{
  "message": "ai.completion.error",
  "properties": {
    "error": "AI.JSX(1032): OpenAI API Error: 404 The model \`unsupported-model\` does not exist.

It's unclear whether this was caused by a bug in AI.JSX, in your code, or is an expected runtime error.
    
Need help? 
* Discord: https://discord.com/channels/1065011484125569147/1121125525142904862
* Docs: https://docs.ai-jsx.com/
* GH: https://github.com/fixie-ai/ai-jsx/issues",
    "latency": Any<Number>,
    "spanId": Any<String>,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions 1`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are an expert in colors. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
      {
        "content": "What do you get when you mix red with blue?",
        "role": "user",
      },
    ],
    "model": "gpt-3.5-turbo",
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions 2`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "purple",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "promptTracking": {
      "id": "determine-color-combination",
      "templates": [
        {
          "id": "determine-color-combination/system",
          "template": "You are an expert in colors. Always respond with one word in all lowercase letters and no punctuation.",
        },
        {
          "id": "determine-color-combination/user",
          "template": "What do you get when you mix red with {{ get-thing-color }}?",
        },
      ],
    },
    "provider": "openai",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 1,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions 3`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
      {
        "content": "What color is the sky?",
        "role": "user",
      },
    ],
    "model": "gpt-3.5-turbo",
    "parentSpanId": Any<String>,
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions 4`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "blue",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "parentSpanId": Any<String>,
    "promptTracking": {
      "id": "get-thing-color",
      "templates": [
        {
          "id": "get-thing-color/system",
          "template": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        },
        {
          "id": "get-thing-color/user",
          "template": "What color is the {{ thing }}?",
        },
      ],
    },
    "provider": "openai",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 1,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions w/o tracking IDs 1`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are an expert in colors. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
      {
        "content": "What do you get when you mix red with blue?",
        "role": "user",
      },
    ],
    "model": "gpt-3.5-turbo",
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions w/o tracking IDs 2`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "purple",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "provider": "openai",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 1,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions w/o tracking IDs 3`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
      {
        "content": "What color is the sky?",
        "role": "user",
      },
    ],
    "model": "gpt-3.5-turbo",
    "parentSpanId": Any<String>,
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles nested completions w/o tracking IDs 4`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "blue",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "parentSpanId": Any<String>,
    "provider": "openai",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 1,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles placeholders 1`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
      {
        "content": "What color is the sky?",
        "role": "user",
      },
    ],
    "model": "gpt-3.5-turbo",
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx handles placeholders 2`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "blue",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "promptTracking": {
      "id": "my-tracker-id",
      "templates": [
        {
          "id": "my-tracker-id/system",
          "template": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        },
        {
          "id": "my-tracker-id/user",
          "template": "What color is the {{ thing }}?",
        },
      ],
    },
    "provider": "openai",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 1,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx works with no tracker id 1`] = `
{
  "message": "ai.completion.request",
  "properties": {
    "messages": [
      {
        "content": "You are a helpful assistant. Always respond with one word in all lowercase letters and no punctuation.",
        "role": "system",
      },
      {
        "content": "What color is the sky?",
        "role": "user",
      },
    ],
    "model": "gpt-3.5-turbo",
    "provider": "openai",
    "spanId": Any<String>,
    "temperature": 0,
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;

exports[`ai-jsx works with no tracker id 2`] = `
{
  "message": "ai.completion.response",
  "properties": {
    "choices": [
      {
        "message": {
          "content": "blue",
          "role": "assistant",
        },
      },
    ],
    "latency": Any<Number>,
    "provider": "openai",
    "spanId": Any<String>,
    "usage": {
      "completion_tokens": 1,
    },
  },
  "timestamp": Any<String>,
  "traceId": Any<String>,
}
`;
